src/rpc/rpc.cc

```diff
@@ -813,8 +814,78 @@ namespace rpc {
       });
   }
 
+  struct client::metrics::domain {
+      metrics::domain_list_t list;
+      stats dead;
+      seastar::metrics::metric_groups metric_groups;
+
+      static thread_local std::unordered_map<sstring, domain> all;
+      static domain& find_or_create(sstring name);
+
+      stats::counter_type count_all(stats::counter_type stats::* field) noexcept {
+          stats::counter_type res = dead.*field;
+          for (const auto& m : list) {
+              res += m._c._stats.*field;
+          }
+          return res;
+      }
+
+      size_t count_all_fn(size_t (client::*fn)(void) const) noexcept {
+          size_t res = 0;
+          for (const auto& m : list) {
```

Comment:
> is the loop here goes over all clients?

All in that domain, yes

> could that be in the range of thousends?

Hardly. It's RPC metrics, so there will be as many sockets as there are nodes in the cluster (minus self) times the number of sockets in domain. The plan is to make one domain per scylla tenant, maybe join some not very interesting domains, so we'll likely have here number of nodes times ~4 sockets. That's not thousands for sure

> Could you hold one counter, gauge for domain per shard, and referense to it from each of the clients on each shard.

As you told, it won't work for queue sizes. Also I explicitly wanted to avoid bumping two counters in two distinct cache-lines upon each RPC operation. Taking into account we're not going to have A LOT of sockets in each domain, looping through them collection-time should be a good trade-off.

Reply:
@amnonh , does this sound reasonable to you? :point_up:

Reply:
@xemul it's reasonable, I just hope that it will not create stalls, as this code runs without preemption, I would have preferred unified counters when possible, it adds some complexity to the code, but you get the sum-part for free.
But maybe it's a premature optimization, and there is no real problem.

Do we have a simple way of turning this off? (In case someone in the field finds a scenario that this implementation starts causing stalls?)
From my point of view if you can identify all those metrics with a label it should be fine to disable it using relabel config.

Reply:
The `domain=foo` can be used I suppose. All sockets get into `domain=default` by default, but scylla is going to create custom domains. Maybe it's worth adding a "shadow" domain that will not register is metrics. Then to turn it off scylla will need an option to create all sockets in shadow domain. WDYT?

Reply:
If the metrics is easy to disable - e.g. using the relabel config or api, or adding a configuration option to skip the registration all together, there's little overhead

Reply:
Can relabel config reference a metrics to be disabled by its name?

Reply:
> Can relabel config reference a metrics to be disabled by its name?

It can, but you have a banch of mertrics, so you want an easy way to disable them all

Reply:
They call it "regular expression" I believe :) Can I disable `.*_rpc_client_.*`-named metrics in one go?

Reply:
yes, you can, maybe it's enough and I'm worried too much

Reply:
Hm, it looks like it can. `s << ((l == "__name__") ? info.id.full_name() : val->second);` is included into the regex matcher.

Reply:
I know that it can. It's just not very intuitive, but as my previous comment, maybe it's not really an issue and I'm worried too much.